### PR TITLE
Unified driver single walker per crowd QMCHamiltonian::flex_evaluate optimization

### PR DESCRIPTION
**Pull request by PDoakORNL**
Restored a working single walker path to the flex_evaluate call.  Checked that RVO actually occurs for returned vector, it does at least with clang 8.  Added force_rvo template which can be used as shown in 
commit 2d77d00

### CODE REVIEW

**Comment by PDoakORNL** [comment]
Also put ContextForSteps on a diet. Going to try having a per walker RandomGen, if that works sensibly class will be eliminated.

**Comment by ye-luo** [comment]
@PDoakORNL I'd like to keep things simple. I believe most compilers have RVO optimization. If not, the extra cost is very low because, the call counts of Hamiltonian evaluate is quite low.

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 1]
```diff
@@ -532,16 +532,16 @@ std::vector<QMCHamiltonian::RealType> QMCHamiltonian::flex_evaluate(const RefVec
...
-    
-    //}
-  // else if (H_list.size() == 1)
-  // {
-  //   H_list[0].get().evaluate(P_list[0]);
-  // }
+  }
+  else if (H_list.size() == 1)
+  {
+    std::vector<RealType> local_energies(1,0.0);
```
Comment (ye-luo) [inline src/QMCHamiltonians/QMCHamiltonian.cpp]: Need resize, not declaration.

Comment (ye-luo) [reply src/QMCHamiltonians/QMCHamiltonian.cpp]: There is resize above. Seems need nothing here.

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 2]
```diff
@@ -491,8 +490,9 @@ QMCHamiltonian::Return_t QMCHamiltonian::evaluate(ParticleSet& P)
 std::vector<QMCHamiltonian::RealType> QMCHamiltonian::flex_evaluate(const RefVector<QMCHamiltonian>& H_list,
                                    const RefVector<ParticleSet>& P_list)
 {
-  // if (H_list.size() > 1)
-  // {
+  std::vector<RealType> local_energies(H_list.size(),0.0);
```
Comment (ye-luo) [inline src/QMCHamiltonians/QMCHamiltonian.cpp]: Although I hate QMCHamiltonian::Return_t, Please use consistent data type as single evaluation.
Return_t is always on full precision. We want local energy not losing accuracy.

File: src/QMCHamiltonians/QMCHamiltonian.cpp (hunk 1 above)
Comment (PDoakORNL) [reply src/QMCHamiltonians/QMCHamiltonian.cpp]: 😢

**Pull request merged**